programs/mango-v4/src/instructions/account_buyback_fees_with_mngo.rs

```diff
@@ -0,0 +1,151 @@
+
+    let bonus_rate = I80F48::from_num(group.fees_mngo_bonus_factor);
+    let now_ts = Clock::get()?.unix_timestamp.try_into().unwrap();
+
+    // quick return if nothing to buyback
+    let mut max_buyback =
+        I80F48::from_num::<u64>(max_buyback.min(account.fixed.discount_buyback_fees_accrued));
+    if max_buyback == I80F48::ZERO {
+        msg!(
+            "nothing to buyback, (discount_buyback_fees_accrued {})",
+            account.fixed.discount_buyback_fees_accrued
+        );
+        return Ok(());
+    }
+
+    // if mngo token position has borrows, skip buyback
+    let (account_mngo_token_position, account_mngo_raw_token_index, _) =
+        account.ensure_token_position(mngo_bank.token_index)?;
+    let account_mngo_native = account_mngo_token_position.native(&mngo_bank);
+    if account_mngo_native.is_negative() {
+        msg!(
+            "account mngo token position ({} native mngo) has borrows, nothing will be bought back",
+            account_mngo_native
+        );
+        return Ok(());
+    }
+
+    // compute max mngo to swap for fees
+    let mngo_oracle_price = mngo_bank.oracle_price(
+        &AccountInfoRef::borrow(&ctx.accounts.mngo_oracle.as_ref())?,
+        Some(Clock::get()?.slot),
+    )?;
+    let mngo_buyback_price = cm!(mngo_oracle_price * bonus_rate);
+    // mngo is exchanged at a discount
+    let mut max_buyback_mngo = cm!(max_buyback / mngo_buyback_price);
+    // buyback is restricted to accounts token position
+    max_buyback_mngo = max_buyback_mngo.min(account_mngo_native);
+    max_buyback = cm!(max_buyback_mngo * mngo_buyback_price);
+
+    // move mngo from user to dao
+    let (dao_mngo_token_position, dao_mngo_raw_token_index, _) =
+        dao_account.ensure_token_position(mngo_bank.token_index)?;
+    let in_use = mngo_bank.withdraw_without_fee(
+        account_mngo_token_position,
+        max_buyback_mngo,
+        now_ts,
+        mngo_oracle_price,
+    )?;
+    if !in_use {
+        account.deactivate_token_position_and_log(
+            account_mngo_raw_token_index,
+            ctx.accounts.account.key(),
+        );
+    }
+    let in_use = mngo_bank.deposit(dao_mngo_token_position, max_buyback_mngo, now_ts)?;
+    if !in_use {
+        dao_account.deactivate_token_position_and_log(
+            dao_mngo_raw_token_index,
+            ctx.accounts.dao_account.key(),
+        );
+    }
+
+    // move fees from dao to user
+    let (account_fees_token_position, account_fees_raw_token_index, _) =
+        account.ensure_token_position(fees_bank.token_index)?;
+    let (dao_fees_token_position, dao_fees_raw_token_index, _) =
+        dao_account.ensure_token_position(fees_bank.token_index)?;
+    let dao_fees_native = dao_fees_token_position.native(&fees_bank);
+    if dao_fees_native.is_negative() || dao_fees_native < max_buyback {
+        msg!(
+            "dao fees token position ({} native fees) is lesser than max buyback {}, nothing will be bought back",
```

Comment:
see comment above, I think the buyback should always be limited to dao_fees_native